### PR TITLE
Add libpcap package

### PR DESCRIPTION
**Pull request by mathbunnyru**
Specify library name and version:  **libpcap/1.9.1**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

This is my first attempt to create a package in CCI. I will try my best to fix everything.
This work is based on https://github.com/bincrafters/conan-libpcap with several fixes.
I will try to explain them.

### CODE REVIEW

File: recipes/libpcap/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+from conans import AutoToolsBuildEnvironment, tools, ConanFile
+from conans.errors import ConanInvalidConfiguration
+import os
+
+
+class LibPcapConan(ConanFile):
+    name = "libpcap"
+    version = "1.9.1"
```
Comment (mathbunnyru) [inline recipes/libpcap/all/conanfile.py:8]: If I do delete this I can't run: conan create all libpcap/1.9.1
But for sure it shouldn't be present here.

File: recipes/libpcap/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        return self.settings.arch == "x86" and tools.detected_architecture() == "x86_64"
+
+    def requirements(self):
+        if self.options.enable_usb:
+            self.requires("libusb/1.0.23")
+
+    def build_requirements(self):
+        if self.settings.os == "Linux":
+            if not tools.which("bison"):
+                self.build_requires("bison/3.7.1")
```
Comment (mathbunnyru) [inline recipes/libpcap/all/conanfile.py:47]: Updated to latest version

File: recipes/libpcap/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            package_list = []
+            if self.options.enable_dbus:
+                package_list.extend(["libdbus-glib-1-dev%s" % arch, "libdbus-1-dev"])
+            if self.options.enable_bluetooth:
+                package_list.append("libbluetooth-dev%s" % arch)
+            if self.options.enable_packet_ring:
+                package_list.append("libnl-genl-3-dev%s" % arch)
+            if package_list:
+                package_tool = tools.SystemPackageTool(conanfile=self)
+                package_tool.install(packages=" ".join(package_list), update=True)
```
Comment (mathbunnyru) [inline recipes/libpcap/all/conanfile.py:63]: I get an error:
[HOOK - conan-center.py] pre_export(): ERROR: [SYSTEM REQUIREMENTS (KB-H032)] The method 'SystemPackageTool.install' is not allowed in the recipe. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H032)

How should I address it?

File: recipes/libpcap/all/conanfile.py (hunk 1 above)
Comment (ericLemanissier) [reply recipes/libpcap/all/conanfile.py:8]: `conan create all libpcap/1.9.1@` should work

File: recipes/libpcap/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_folder = self.name + "-" + self.name + "-" + self.version
+        os.rename(extracted_folder, self._source_subfolder)
+
+    def configure(self):
+        if self.settings.os == "Windows":
+            raise ConanInvalidConfiguration("For Windows use winpcap/4.1.2@bincrafters/stable")
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
```
Comment (mathbunnyru) [inline recipes/libpcap/all/conanfile.py:74]: Pure C library

File: recipes/libpcap/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    def _is_amd64_to_i386(self):
+        return self.settings.arch == "x86" and tools.detected_architecture() == "x86_64"
+
+    def requirements(self):
+        if self.options.enable_usb:
+            self.requires("libusb/1.0.23")
+
+    def build_requirements(self):
+        if self.settings.os == "Linux":
+            if not tools.which("bison"):
```
Comment (ericLemanissier) [inline recipes/libpcap/all/conanfile.py:46]: we usually remove this if, in order to use the same bison version everywhere (same for flex below)

File: recipes/libpcap/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        if self.settings.os == "Windows":
+            raise ConanInvalidConfiguration("For Windows use winpcap/4.1.2@bincrafters/stable")
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def _configure_autotools(self):
+        if not self._autotools:
+            self._autotools = AutoToolsBuildEnvironment(self)
+            configure_args = ["--prefix=%s" % self.package_folder]
+            configure_args.append("--enable-shared" if self.options.shared else "--disable-shared")
```
Comment (mathbunnyru) [inline recipes/libpcap/all/conanfile.py:80]: Deleted a line with --enable-static (--disable-static) here, but it doesn't change anything and gives a warning.

File: recipes/libpcap/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    def build(self):
+        autotools = self._configure_autotools()
+        autotools.make()
+
+    def package(self):
+        self.copy("LICENSE", src=self._source_subfolder, dst="licenses")
+        autotools = self._configure_autotools()
+        autotools.install()
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
```
Comment (mathbunnyru) [inline recipes/libpcap/all/conanfile.py:105]: There is a .pc file here, so added this line

File: recipes/libpcap/all/test_package/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.1)
+project(libpcap_example)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(list_network list_network.cpp)
+target_link_libraries(list_network ${CONAN_LIBS})
```
Comment (mathbunnyru) [inline recipes/libpcap/all/test_package/CMakeLists.txt:8]: Changed some names to make them more user-friendly

**Comment by mathbunnyru** [comment]
I did run this under my macOS and under Ubuntu 18.04
Seems to work.

File: recipes/libpcap/all/conanfile.py (hunk 3 above)
Comment (ericLemanissier) [reply recipes/libpcap/all/conanfile.py:63]: you should either create a regular conan package for these requirements, or create a system package for each of these requirements (in the same vein as https://github.com/conan-io/conan-center-index/blob/master/recipes/opengl/all/conanfile.py), or remove the corresponding option

File: recipes/libpcap/all/conanfile.py (hunk 1 above)
Comment (mathbunnyru) [reply recipes/libpcap/all/conanfile.py:8]: That wasn't obvious at all, thanks!

File: recipes/libpcap/all/conanfile.py (hunk 3 above)
Comment (mathbunnyru) [reply recipes/libpcap/all/conanfile.py:60]: What would you recommend to do?
Removing the options seems to be quite easy, but I'm not sure how useful they are.

Comment (ericLemanissier) [reply recipes/libpcap/all/conanfile.py:60]: If the pcap is usable without these options, I guess you can just remove them, and wait until someone raises an issue, or makes a pull request.
DISCLAIMER: I don't know this library at all

Comment (uilianries) [reply recipes/libpcap/all/conanfile.py:60]: I made the original version for Bincrafters. pcap is used to analyze data packages, (e.g. networking). My Recommendation is removing those options and add a TODO note in this recipe:

# TODO: Add dbus-glib when available
# TODO: Add libbluetooth when available
# TODO: Add libnl-genl when available

**Comment by mathbunnyru** [comment]
Thanks, @uilianries !
Sorry for taking some time to response, I was on a small vacation :)

**Comment by mathbunnyru** [comment]
Checked on my macOS, seems to work

File: recipes/libpcap/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    # TODO: Add dbus-glib when available
+    # TODO: Add libnl-genl when available
+    # TODO: Add libbluetooth when available
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def requirements(self):
```
Comment (uilianries) [inline recipes/libpcap/all/conanfile.py:36]:
```suggestion
    def configure(self):
        if self.settings.os == "Macos" and self.options.shared:
            raise ConanInvalidConfiguration("libpcap can not be built as shared on OSX.")

    def requirements(self):
```

File: recipes/libpcap/all/test_package/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.1)
+project(libpcap_example)
```
Comment (Croydon) [inline recipes/libpcap/all/test_package/CMakeLists.txt:2]:
```suggestion
project(test_package)
```

Please leave this generic. Should be the same for all recipes.

File: recipes/libpcap/all/test_package/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        bin_path = os.path.join("bin", "list_network")
```
Comment (Croydon) [inline recipes/libpcap/all/test_package/conanfile.py:15]:
```suggestion
        bin_path = os.path.join("bin", "test_package")
```

**Comment by mathbunnyru** [comment]
@Croydon I also had to rename executable after your suggestion.
I don't know what is preferred - common name "test_package" or something more specific

**Comment by mathbunnyru** [comment]
Should be good to be merged 🙏

**Review by jgsogo — COMMENTED** [review]
Thanks @mathbunnyru for your first package of many to come 😄 

I really hope your experience adding the recipe and the automatic reviews (and the invaluable feedback from the community) has helped you. Any feedback about the process or things you think we should improve are more than welcome...

**Review by jgsogo — APPROVED** [review]

**Comment by mathbunnyru** [comment]
Thanks @jgsogo !

I did like the experience and I think we (me and my colleagues) will add several packages in the nearest future.
I also liked the community response - it's awesome.

The things I didn't like:
1. Unexpected errors. I think it would be great to have some logs when this happens.
2. Sometimes I need to merge the master and restart the build - it's also a bit disappointing.

**Comment by prince-chrismc** [comment]
The CCI team is very active, they restart jobs that unexpectedly failed. You are not obligated to do so :smile: 

Since their pipeline is still private, I suspect they do not post logs to avoid leaking business secrets. (unfortunately for us business users)

Welcome to the contributors list!

**Comment by jgsogo** [comment]
Thanks for the feedback!

We are working hard in the pipeline (but there are many things to work on, it is not only Conan Center). We would really like to improve the information about those unexpected errors, usually it's some failure unrelated to the recipe itself: timeouts, connection errors,... we will add some retries to alleviate it (maybe a bot that triggers the retry for unexpected failures), and see if we can capture more information to show here in Github.

About the pipeline being private, we really think it is very ConanCenter specific and it won't be very useful for general usage... but we are learning a lot from it. In the future, we might consider splitting it into two different libraries and open-source everything that could be used by other pipelines. Nevertheless, any question about the CI is also welcome and we will try to answer it the best we can (there is nothing _secret_).

@mathbunnyru I don't know what you mean when you say _you need to merge the master_, PRs don't need to be up-to-date with master before merging them... is it just about triggering again the CI for unexpected failures?

**Comment by mathbunnyru** [comment]
Yes, it was just about triggering CI once again.

**Pull request merged**